src/util-interval-tree.c

```diff
@@ -0,0 +1,421 @@
+                    a->max, b->port, b->port2, b->max);
+        }
+        SCLogDebug("a: [%u:%u, max %u] b: [%u:%u, max %u] => 1", a->port, a->port2, a->max, b->port,
+                b->port2, b->max);
+        SCReturnInt(1);
+    }
+    SCLogDebug("a: [%u:%u, max %u] b: [%u:%u, max %u] => -1", a->port, a->port2, a->max, b->port,
+            b->port2, b->max);
+    SCReturnInt(-1);
+}
+
+IRB_GENERATE(PI, SCIntervalNode, irb, PICompareAndUpdate);
+
+/**
+ * \brief Function to initialize the interval tree.
+ *
+ * \return Pointer to the newly created interval tree
+ */
+SCIntervalTree *SCIntervalTreeInit(void)
+{
+    SCIntervalTree *it = SCCalloc(1, sizeof(SCIntervalTree));
+    if (it == NULL) {
+        return NULL;
+    }
+
+    return it;
+}
+
+/**
+ * \brief Helper function to free a given node in the interval tree.
+ *
+ * \param de_ctx Detection Engine Context
+ * \param it Pointer to the interval tree
+ */
+static void SCIntervalNodeFree(DetectEngineCtx *de_ctx, SCIntervalTree *it)
+{
+    SCIntervalNode *node = NULL, *safe = NULL;
+    IRB_FOREACH_SAFE(node, PI, &it->tree, safe)
+    {
+        SigGroupHeadFree(de_ctx, node->sh);
+        PI_IRB_REMOVE(&it->tree, node);
+        SCFree(node);
+    }
+    it->head = NULL;
+}
+
+/**
+ * \brief Function to free an entire interval tree.
+ *
+ * \param de_ctx Detection Engine Context
+ * \param it Pointer to the interval tree
+ */
+void SCIntervalTreeFree(DetectEngineCtx *de_ctx, SCIntervalTree *it)
+{
+    if (it) {
+        SCIntervalNodeFree(de_ctx, it);
+        SCFree(it);
+    }
+}
+
+/**
+ * \brief Function to insert a node in the interval tree.
+ *
+ * \param de_ctx Detection Engine Context
+ * \param it Pointer to the interval tree
+ * \param head Pointer to the head of the tree named PI
+ * \param p Pointer to a DetectPort object
+ *
+ * \return SC_OK if the node was inserted successfully, SC_EINVAL otherwise
+ */
+int PIInsertPort(DetectEngineCtx *de_ctx, SCIntervalTree *it, struct PI *head, const DetectPort *p)
+{
+    DEBUG_VALIDATE_BUG_ON(p->port > p->port2);
+
+    SCIntervalNode *pi = SCCalloc(1, sizeof(*pi));
+    if (pi == NULL) {
+        return SC_EINVAL;
+    }
+
+    pi->port = p->port;
+    pi->port2 = p->port2;
+    pi->flags = p->flags;
+    SigGroupHeadCopySigs(de_ctx, p->sh, &pi->sh);
+
+    if (PI_IRB_INSERT(&it->tree, pi) != NULL) {
+        SCLogDebug("Node wasn't added to the tree: port: %d, port2: %d", pi->port, pi->port2);
+        SCFree(pi);
+        return SC_EINVAL;
+    }
+    return SC_OK;
+}
+
+/**
+ * \brief Function to check if a port range overlaps with a given set of ports
+ *
+ * \param port Given low port
+ * \param port2 Given high port
+ * \param ptr Pointer to the node in the tree to be checked against
+ *
+ * \return true if an overlaps was found, false otherwise
+ */
+static bool IsOverlap(const uint16_t port, const uint16_t port2, const SCIntervalNode *ptr)
```

Comment:
needs careful review

Reply:
one thing that I find counter intuitive about the current code is how it creates the ranges wrt borders: 0-80, 80-81, 81-100. Where does a port range 81 fall in?

In the old code it would create non-overlapping ranges, e.g. 0-79, 80-81, 82-100.

Reply:
This has been fixed in the latest revs. Now the ports are only seen once.